File: functions/scripts/python/story_generator.py
"""
Core functionality for generating stories using OpenAI and LangGraph.
"""

import os
from dotenv import load_dotenv
from typing import List, Dict, Tuple
from openai import OpenAI
from langgraph.func import task, entrypoint
from langgraph.checkpoint.memory import MemorySaver
import logging
import time
import textwrap
from pydantic import BaseModel
from prompts import (
    SCRIPT_GENERATION_PROMPT,
    KEYFRAME_EXTRACTION_PROMPT,
    DIALOG_EXTRACTION_PROMPT,
    VISUAL_STYLE_PROMPT
)
from enum import Enum
from tts_core import generate_speech_from_text

class LeonardoStyles(str, Enum):
    RENDER_3D = "RENDER_3D"
    ACRYLIC = "ACRYLIC"
    ANIME_GENERAL = "ANIME_GENERAL"
    CREATIVE = "CREATIVE"
    DYNAMIC = "DYNAMIC"
    FASHION = "FASHION"
    GAME_CONCEPT = "GAME_CONCEPT"
    GRAPHIC_DESIGN_3D = "GRAPHIC_DESIGN_3D"
    ILLUSTRATION = "ILLUSTRATION"
    NONE = "NONE"
    PORTRAIT = "PORTRAIT"
    PORTRAIT_CINEMATIC = "PORTRAIT_CINEMATIC"
    RAY_TRACED = "RAY_TRACED"
    STOCK_PHOTO = "STOCK_PHOTO"
    WATERCOLOR = "WATERCOLOR"

class Character(str, Enum):
    NARRATOR = "narrator"
    CHILD = "child"
    ELDER = "elder"
    FAIRY = "fairy"
    HERO = "hero"
    VILLAIN = "villain"


class VisualStyleResponse(BaseModel):
    style: LeonardoStyles

class DialogResponse(BaseModel):
    character: Character
    text: str

# Set up logging configuration
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S"
)
logger = logging.getLogger(__name__)

# Load environment variables from specified path
env_path = os.path.join(os.path.dirname(__file__), "../../.env")
load_dotenv(env_path)

# Initialize OpenAI client
client = OpenAI()

# Define response schema classes
class Keyframe(BaseModel):
    title: str
    description: str

class KeyframeResponse(BaseModel):
    keyframes: List[Keyframe]

# ---------- Helper Functions ----------

def log_prompt_and_response(prompt: str, response: str, context: str = ""):
    """Helper function to log prompts and responses in a readable format."""
    if context:
        logger.debug("\n=== %s ===", context)
    logger.debug("\n----- Prompt -----\n%s", textwrap.indent(prompt, "> "))
    logger.debug("\n----- Response -----\n%s", textwrap.indent(response, "< "))
    logger.debug("-" * 60)

def invoke_chat_completion(prompt: str, response_format=None):
    """
    Helper function to invoke OpenAI chat completion API.

    Args:
        prompt: The prompt to send to the API.
        response_format: Optional Pydantic class specifying the response format (for structured outputs).

    Returns:
        If response_format is provided, returns the response object from the beta.parse() method.
        Otherwise, returns the message object from a standard completion.
    """
    params = {
        "model": "gpt-4o-mini",
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.7,
    }
    if response_format:
        # Use the beta parsing method which enforces structured outputs based on the provided Pydantic model.
        response = client.beta.chat.completions.parse(
            model=params["model"],
            messages=params["messages"],
            temperature=params["temperature"],
            response_format=response_format,
        )
        return response
    else:
        response = client.chat.completions.create(**params)
        return response.choices[0].message

# ---------- OpenAI Pipeline Tasks using LangGraph ----------

@task
def generate_script(keywords: List[str]) -> str:
    """
    Generates a full screenplay script from a list of theme keywords.
    """
    logger.info("Generating script with keywords: %s", ", ".join(keywords))
    start_time = time.time()
    
    prompt = SCRIPT_GENERATION_PROMPT.format(keywords=", ".join(keywords))
    logger.debug("Sending prompt to OpenAI (length: %d characters)", len(prompt))
    
    response = invoke_chat_completion(prompt)
    # For standard completions, response is a message object with a 'content' attribute.
    script = response.content if hasattr(response, "content") else response

    elapsed_time = time.time() - start_time
    logger.info("Script generation completed in %.2f seconds", elapsed_time)
    logger.debug("Generated script length: %d characters", len(script))
    
    log_prompt_and_response(prompt, script, "Script Generation")
    return script

@task
def extract_keyframes(script: str) -> List[Dict[str, str]]:
    """
    Breaks the screenplay into key visual moments ("keyframes") optimized for prompting an image generator.
    Returns the keyframes directly as a list of dictionaries.
    """
    logger.info("Extracting keyframes from script")
    start_time = time.time()
    
    prompt = KEYFRAME_EXTRACTION_PROMPT.format(script=script)
    logger.debug("Sending keyframe extraction prompt to OpenAI")
    
    # Use beta parsing with our Pydantic schema.
    response = invoke_chat_completion(prompt, response_format=KeyframeResponse)
    # Access the parsed result from the beta response.
    parsed_message = response.choices[0].message.parsed
    keyframes = parsed_message.keyframes
    log_prompt_and_response(prompt, str(parsed_message), "Keyframe Extraction")
    
    elapsed_time = time.time() - start_time
    logger.info("Extracted %d keyframes in %.2f seconds", len(keyframes), elapsed_time)
    
    # Convert each Pydantic model instance to a dict.
    return [kf.model_dump() for kf in keyframes]

@task
def extract_dialog(script: str, keyframe_description: str) -> Tuple[Character, str]:
    """
    Combines the full screenplay and a keyframe description to extract dialogue and narration.
    Returns a tuple of (character, text).
    """
    logger.debug("Extracting dialog for keyframe: %s", keyframe_description[:100] + "...")
    start_time = time.time()
    
    prompt = DIALOG_EXTRACTION_PROMPT.format(
        keyframe_description=keyframe_description,
        script=script
    )
    response = invoke_chat_completion(prompt, response_format=DialogResponse)
    dialog = response.choices[0].message.parsed
    
    # Create a JSON-compatible dict for logging
    dialog_dict = {"character": dialog.character, "text": dialog.text}
    log_prompt_and_response(prompt, str(dialog_dict), 
                          f"Dialog Extraction (Keyframe: {keyframe_description[:50]}...)")
    
    elapsed_time = time.time() - start_time
    logger.debug("Dialog extraction completed in %.2f seconds", elapsed_time)
    
    return dialog.character, dialog.text

@task
def determine_visual_style(script: str) -> LeonardoStyles:
    """
    Analyzes the script and determines the most appropriate visual style for the scenes.
    Returns one of the predefined LeonardoStyles.
    """
    logger.info("Determining visual style for the script")
    start_time = time.time()
    
    prompt = VISUAL_STYLE_PROMPT.format(script=script)
    logger.debug("Sending visual style prompt to OpenAI")
    
    response = invoke_chat_completion(prompt, response_format=VisualStyleResponse)
    # Access the parsed result from the beta response
    style = response.choices[0].message.parsed.style
    
    log_prompt_and_response(prompt, str(style), "Visual Style Determination")
    
    elapsed_time = time.time() - start_time
    logger.info("Visual style determined in %.2f seconds: %s", elapsed_time, style)
    
    return style

@task
def generate_voiceover(character: Character, text: str) -> bytes:
    """
    Generates audio voiceover for the given character and text using TTS.
    
    Args:
        character: The character type (narrator, child, etc.)
        text: The text to convert to speech
        
    Returns:
        Audio data as bytes
    """
    logger.info(f"Generating voiceover for {character}: {text[:50]}...")
    start_time = time.time()
    
    try:
        response = generate_speech_from_text(text, character=character.value)
        audio_data = response.read()
        
        elapsed_time = time.time() - start_time
        logger.info(f"Voiceover generation completed in {elapsed_time:.2f} seconds")
        
        return audio_data
        
    except Exception as e:
        logger.error(f"Failed to generate voiceover: {str(e)}")
        raise

@entrypoint(checkpointer=MemorySaver())
def generate_story(inputs: Dict[str, List[str]]) -> List[Tuple[str, Tuple[Character, str], str, bytes]]:
    """
    Public interface function that ties the OpenAI steps together.

    Input:
      - inputs: Dictionary containing 'keywords' list

    Output:
      - A list of tuples, where each tuple is (keyframe description, keyframe dialog/narration, visual style, audio_data).
    """
    logger.info("Starting story generation process")
    start_time = time.time()
    
    keywords = inputs["keywords"]
    logger.info("Processing keywords: %s", ", ".join(keywords))
    
    # Generate the initial script
    script = generate_script(keywords).result()
    
    # Parallel tasks: Extract keyframes and determine visual style
    keyframes = extract_keyframes(script).result()
    visual_style = determine_visual_style(script).result()
    
    # First, launch all dialog extraction tasks in parallel
    logger.info("Launching parallel dialog extraction for %d keyframes", len(keyframes))
    dialog_tasks = [extract_dialog(script, kf["description"]) for kf in keyframes]
    
    # Wait for all dialogs to complete
    dialogs = [task.result() for task in dialog_tasks]
    
    # Now launch all voiceover generation tasks in parallel
    logger.info("Launching parallel voiceover generation for %d keyframes", len(keyframes))
    voiceover_tasks = [generate_voiceover(character, text) for character, text in dialogs]
    
    # Wait for all voiceovers to complete
    voiceovers = [task.result() for task in voiceover_tasks]
    
    # Combine results
    results = [
        (kf["description"], dialog, visual_style, voiceover)
        for kf, dialog, voiceover in zip(keyframes, dialogs, voiceovers)
    ]
    
    total_time = time.time() - start_time
    logger.info("Story generation completed in %.2f seconds", total_time)
    
    return results 

File: functions/src/story_generator.ts
import {OpenAI} from 'openai';
import {task, entrypoint, MemorySaver} from '@langchain/langgraph';
import type {LangGraphRunnableConfig} from '@langchain/langgraph';
import * as path from 'path';
import * as fs from 'fs/promises';
import {z} from 'zod';
import {zodResponseFormat} from 'openai/helpers/zod';
import {openAiKey} from './config.js';
import {
  type Character,
  type Characters,
  type VisualStyleResponse,
  type Keyframe,
  type KeyframeResponse,
  type KeyframeScene,
  type KeyframesWithDialog,
  type LeonardoStyle
} from './schemas.js';
import {generateSpeechFromText} from './tts_core.js';
import {
  SCRIPT_GENERATION_PROMPT,
  KEYFRAME_EXTRACTION_PROMPT,
  VISUAL_STYLE_PROMPT,
  CHARACTER_EXTRACTION_PROMPT,
  SCRIPT_ENHANCEMENT_PROMPT,
  LEONARDO_PROMPT_OPTIMIZATION,
  WORDS_PER_KEYFRAME,
  KEYFRAME_SCENE_GENERATION_PROMPT
} from './prompts.js';
import {uploadAudioToMux} from './mux_core.js';

// Initialize OpenAI client
let client: OpenAI;

// Helper function to get OpenAI client (initialized at runtime)
function getOpenAIClient(): OpenAI {
  if (!client) {
    client = new OpenAI({
      apiKey: openAiKey.value()
    });
  }
  return client;
}

// Helper function to invoke OpenAI chat completion
async function invokeChatCompletion<T>(
  prompt: string,
  responseFormat?: { schema: z.ZodSchema<T> }
): Promise<T | string> {
  const params = {
    model: "gpt-4o-mini",
    messages: [{ role: "user" as const, content: prompt }],
    temperature: 0.7,
  };

  if (responseFormat) {
    // Get the schema description or fall back to a generic name
    const schemaDescription = responseFormat.schema.description || 'structured_response';

    const completion = await getOpenAIClient().beta.chat.completions.parse({
      ...params,
      response_format: zodResponseFormat(responseFormat.schema, schemaDescription)
    });

    const message = completion.choices[0]?.message;
    if (!message?.parsed) {
      throw new Error("No parsed response received");
    }
    return message.parsed as T;
  } else {
    const response = await getOpenAIClient().chat.completions.create(params);
    return response.choices[0].message.content!;
  }
}

// Define Zod schemas for our types
const KeyframeResponseSchema = z.object({
  keyframes: z.array(z.object({
    title: z.string(),
    description: z.string(),
    characters_in_scene: z.array(z.string())
  }))
});

const VisualStyleResponseSchema = z.object({
  style: z.enum([
    "RENDER_3D", "ACRYLIC", "ANIME_GENERAL", "CREATIVE", "DYNAMIC",
    "FASHION", "GAME_CONCEPT", "GRAPHIC_DESIGN_3D", "ILLUSTRATION",
    "NONE", "PORTRAIT", "PORTRAIT_CINEMATIC", "RAY_TRACED",
    "STOCK_PHOTO", "WATERCOLOR"
  ])
});

const CharactersSchema = z.object({
  characters: z.array(z.object({
    role: z.enum(["narrator", "child", "elder", "fairy", "hero", "villain", "sage", "sidekick"]),
    name: z.string(),
    backstory: z.string(),
    physical_description: z.string(),
    personality: z.string()
  }))
});

const KeyframesWithDialogSchema = z.object({
  narratedScene: z.object({
    title: z.string(),
    description: z.string(),
    characters_in_scene: z.array(z.string()),
    dialog: z.string(),
    leonardo_prompt: z.string().optional(),
    character: z.object({
      role: z.literal("narrator"),
      name: z.literal("Narrator"),
      backstory: z.literal("I am the eternal storyteller, weaving tales through time and space."),
      physical_description: z.literal("A warm, resonant voice that brings stories to life."),
      personality: z.literal("Wise and engaging, I share stories with wonder and insight, drawing listeners into each magical moment.")
    })
  }),
  dialoguedScene: z.object({
    title: z.string(),
    description: z.string(),
    characters_in_scene: z.array(z.string()),
    dialog: z.string(),
    leonardo_prompt: z.string().optional(),
    character: z.object({
      role: z.enum(["child", "elder", "fairy", "hero", "villain", "sage", "sidekick"]),
      name: z.string(),
      backstory: z.string(),
      physical_description: z.string(),
      personality: z.string()
    })
  })
});

// Task definitions
export const generateScript = task("generate_script", async (keywords: string[]): Promise<string> => {
  console.log("Generating script with keywords:", keywords.join(", "));
  const prompt = SCRIPT_GENERATION_PROMPT.replace("{keywords}", keywords.join(", "));
  return await invokeChatCompletion(prompt) as string;
});

export const extractKeyframes = task("extract_keyframes", async (
  script: string,
  numKeyframes: number
): Promise<KeyframeResponse> => {
  console.log("Extracting keyframes from script");
  const prompt = KEYFRAME_EXTRACTION_PROMPT
    .replace("{script}", script)
    .replace("{num_keyframes}", numKeyframes.toString());
  
  return await invokeChatCompletion(prompt, { schema: KeyframeResponseSchema }) as KeyframeResponse;
});

export const determineVisualStyle = task("determine_visual_style", async (
  script: string
): Promise<LeonardoStyle> => {
  console.log("Determining visual style for the script");
  const prompt = VISUAL_STYLE_PROMPT.replace("{script}", script);
  const response = await invokeChatCompletion(prompt, { schema: VisualStyleResponseSchema }) as VisualStyleResponse;
  return response.style;
});

export const generateVoiceover = task("generate_voiceover", async (
  scene: KeyframeScene
): Promise<KeyframeScene> => {
  console.log(`Generating voiceover for ${scene.character.name} (${scene.character.role}): ${scene.dialog.slice(0, 50)}...`);
  
  // Generate the audio
  const audioBuffer = await generateSpeechFromText(scene.dialog, scene.character.role);
  
  // Upload to Mux
  const {uploadId, url} = await uploadAudioToMux(audioBuffer, {
    filename: `${scene.title.toLowerCase().replace(/\s+/g, '_')}_audio.mp3`,
    description: `Voiceover for scene: ${scene.title}`
  });
  
  // Update the scene with Mux information
  return {
    ...scene,
    audio: {
      muxUploadId: uploadId,
      muxUrl: url
    }
  };
});

export const extractCharacters = task("extract_characters", async (
  script: string
): Promise<Character[]> => {
  console.log("Extracting characters from script");
  const prompt = CHARACTER_EXTRACTION_PROMPT.replace("{script}", script);
  const response = await invokeChatCompletion(prompt, { schema: CharactersSchema }) as Characters;
  return response.characters;
});

export const enhanceScript = task("enhance_script", async (
  script: string,
  characters: Character[],
  numKeyframes: number
): Promise<string> => {
  console.log("Enhancing script with character details");
  
  const charInfo = characters.map(char => 
    `Character: ${char.name} (${char.role})\n` +
    `Backstory: ${char.backstory}\n` +
    `Personality: ${char.personality}`
  ).join("\n\n");
  
  const totalWords = numKeyframes * WORDS_PER_KEYFRAME;
  
  const prompt = SCRIPT_ENHANCEMENT_PROMPT
    .replace("{script}", script)
    .replace("{characters}", charInfo)
    .replace("{total_words}", totalWords.toString());
  
  return await invokeChatCompletion(prompt) as string;
});

export const optimizeLeonardoPrompts = task("optimize_leonardo_prompts", async (
  keyframes: Array<{
    description: string;
    characters_in_scene: string[];
    title: string;
  }>,
  visualStyle: LeonardoStyle,
  outputDir: string
): Promise<string[]> => {
  console.log("Optimizing all scenes for Leonardo prompts");

  const sceneDescriptions = keyframes.map((kf, idx) => 
    `Scene ${idx + 1}:\n` +
    `Title: ${kf.title}\n` +
    `Description: ${kf.description}\n` +
    `Characters Present: ${kf.characters_in_scene.length ? kf.characters_in_scene.join(", ") : "None"}`
  ).join("\n\n");

  const prompt = LEONARDO_PROMPT_OPTIMIZATION
    .replace("{keyframe_descriptions}", sceneDescriptions)
    .replace("{visual_style}", visualStyle);

  const response = await invokeChatCompletion(prompt) as string;
  
  // Parse the numbered responses
  const optimizedPrompts: string[] = [];
  let currentPrompt: string[] = [];
  let currentNumber = 1;

  for (const line of response.split('\n')) {
    const trimmedLine = line.trim();
    if (!trimmedLine) continue;

    if (trimmedLine.startsWith(`${currentNumber}:`)) {
      if (currentPrompt.length) {
        optimizedPrompts.push(currentPrompt.join('\n'));
        currentPrompt = [];
      }
      currentPrompt.push(trimmedLine.slice(String(currentNumber).length + 1).trim());
      currentNumber++;
    } else {
      currentPrompt.push(trimmedLine);
    }
  }

  if (currentPrompt.length) {
    optimizedPrompts.push(currentPrompt.join('\n'));
  }

  // Save prompts to files
  const outputPath = path.join(outputDir);
  await fs.mkdir(outputPath, { recursive: true });

  await Promise.all(optimizedPrompts.map(async (promptText, idx) => {
    const promptFile = path.join(outputPath, `scene_prompt_${idx + 1}.txt`);
    await fs.writeFile(promptFile, promptText);
  }));

  return optimizedPrompts;
});

export const generateKeyframeScenes = task("generate_keyframe_scenes", async (
  keyframe: Keyframe,
  characters: Character[],
  previouslySeenCharacters: Set<string>
): Promise<KeyframesWithDialog> => {
  console.log(`Generating scenes for keyframe: ${keyframe.title}`);

  const characterProfiles = characters.map(char =>
    `Character: ${char.name} (${char.role})\n` +
    `Personality: ${char.personality}\n` +
    `Physical Description: ${char.physical_description}\n` +
    "---"
  ).join("\n");

  const prompt = KEYFRAME_SCENE_GENERATION_PROMPT
    .replace("{title}", keyframe.title)
    .replace("{description}", keyframe.description)
    .replace("{characters_in_scene}", keyframe.characters_in_scene.join(", "))
    .replace("{character_profiles}", characterProfiles);

  const result = await invokeChatCompletion(prompt, { schema: KeyframesWithDialogSchema }) as KeyframesWithDialog;

  // Update the set of previously seen characters
  [result.narratedScene, result.dialoguedScene].forEach(scene => {
    scene.characters_in_scene.forEach(charName => {
      previouslySeenCharacters.add(charName);
    });
  });

  return result;
});

// Main entrypoint
export interface StoryInput {
  keywords: string[];
}

export interface StoryConfig {
  extract_chars?: boolean;
  generate_voiceover?: boolean;
  generate_images?: boolean;
  generate_motion?: boolean;
  save_script?: boolean;
  num_keyframes?: number;
  output_dir?: string;
}

export type StoryResult = [
  Array<[string, KeyframeScene[], LeonardoStyle, ArrayBuffer | null]>,
  Character[] | null,
  [string, string] | null
];

export const generateStory = entrypoint(
  { checkpointer: new MemorySaver(), name: "generate_story" },
  async (
    inputs: StoryInput,
    config?: LangGraphRunnableConfig & { configurable?: StoryConfig }
  ): Promise<StoryResult> => {
    console.log("Starting story generation process");
    const startTime = Date.now();

    const keywords = inputs.keywords;
    console.log("Processing keywords:", keywords.join(", "));

    // Get config options
    const storyConfig = config?.configurable ?? {};
    const extractChars = storyConfig.extract_chars ?? false;
    const shouldGenerateVoiceover = storyConfig.generate_voiceover ?? false;
    const shouldGenerateImages = storyConfig.generate_images ?? false;
    const saveScript = storyConfig.save_script ?? false;
    const numKeyframes = storyConfig.num_keyframes ?? 4;
    const outputDir = storyConfig.output_dir ?? "output";

    // Create output directory
    await fs.mkdir(outputDir, { recursive: true });

    // Generate initial script
    const script = await generateScript(keywords);
    const characters = await extractCharacters(script);
    const enhancedScript = await enhanceScript(script, characters, numKeyframes);

    // Extract keyframes and determine visual style in parallel
    const [keyframeResponse, visualStyle] = await Promise.all([
      extractKeyframes(enhancedScript, numKeyframes),
      determineVisualStyle(enhancedScript)
    ]);

    const keyframes = keyframeResponse.keyframes;
    const previouslySeenCharacters = new Set<string>();

    // Generate scenes for each keyframe
    const allScenes = await Promise.all(
      keyframes.map((kf: Keyframe) => generateKeyframeScenes(kf, characters, previouslySeenCharacters))
    );

    // Handle image generation if enabled
    if (shouldGenerateImages) {
      const sceneDescriptions = allScenes.flatMap((keyframeScenes: KeyframesWithDialog) => [
        {
          description: keyframeScenes.narratedScene.description,
          characters_in_scene: keyframeScenes.narratedScene.characters_in_scene,
          title: keyframeScenes.narratedScene.title
        },
        {
          description: keyframeScenes.dialoguedScene.description,
          characters_in_scene: keyframeScenes.dialoguedScene.characters_in_scene,
          title: keyframeScenes.dialoguedScene.title
        }
      ]);

      const optimizedPrompts = await optimizeLeonardoPrompts(
        sceneDescriptions,
        visualStyle,
        outputDir
      );

      // Update scene descriptions with optimized prompts
      let promptIdx = 0;
      for (const keyframeScenes of allScenes) {
        keyframeScenes.narratedScene.leonardo_prompt = optimizedPrompts[promptIdx++];
        keyframeScenes.dialoguedScene.leonardo_prompt = optimizedPrompts[promptIdx++];
      }
    }

    // Generate voiceovers if requested
    let voiceovers: KeyframeScene[] = [];
    if (shouldGenerateVoiceover) {
      const voiceoverTasks = allScenes.flatMap((keyframeScenes: KeyframesWithDialog) => [
        generateVoiceover(keyframeScenes.narratedScene),
        generateVoiceover(keyframeScenes.dialoguedScene)
      ]);
      voiceovers = await Promise.all(voiceoverTasks);
      
      // Update scenes with voiceover information
      for (let i = 0; i < allScenes.length; i++) {
        const voiceoverIdx = i * 2;
        allScenes[i].narratedScene = voiceovers[voiceoverIdx];
        allScenes[i].dialoguedScene = voiceovers[voiceoverIdx + 1];
      }
    }

    // Combine results
    const storyResults: StoryResult[0] = allScenes.map((scenes: KeyframesWithDialog) => [
      scenes.narratedScene.description,
      [scenes.narratedScene, scenes.dialoguedScene],
      visualStyle,
      null // Remove the old audio buffer since we now store it in Mux
    ]);

    const totalTime = (Date.now() - startTime) / 1000;
    console.log(`Story generation completed in ${totalTime.toFixed(2)} seconds`);

    return [
      storyResults,
      extractChars ? characters : null,
      saveScript ? [script, enhancedScript] : null
    ];
  }
); 